### ISSUE TITLE
Doxygen for MathematicalProgram is unreadable

### ISSUE THREAD

**Issue by ggould-tri** [RobotLocomotion/drake#11317]
Doxygen rendering of return types makes `drake::solvers::MathematicalProgram` documentation nearly unreadable.

Visit https://drake.mit.edu/doxygen_cxx/classdrake_1_1solvers_1_1_mathematical_program.html
Observe that the page doesn't contain any content because everything is off the right margin:

![bad-drake](https://user-images.githubusercontent.com/17596413/56681582-9e185380-6697-11e9-9623-c878f21f6bf5.png)

Note that the layout culprit is the return type of `GetInitialGuess`, which is `std::enable_if< std::is_same< typename Derived::Scalar, symbolic::Variable >::value, Eigen::Matrix< double, Derived::RowsAtCompileTime, Derived::ColsAtCompileTime > >::type` (that's a mouthful!).

The solution might be either to give doxygen a better custom CSS or else to typedef some of the more awful return type subclauses.

**Comment by jwnimmer-tri** [comment]
The solution that comes to my mind is to put ifdef doxygen around the return type, as I did in https://github.com/RobotLocomotion/drake/blob/2b41c1b655acd2fa86902b84fc53e8b9857cd6a2/systems/framework/input_port.h#L68.  It's probably going to be difficult to get typedefs to work.

**Comment by jwnimmer-tri** [comment]
Now that we have a modern C++, another possible solution would be to rewrite the methods to use `requires` syntax instead of `enable_if`.

**Comment by j4yyousi** [comment]
Hi, here are the before and after pictures showing the effect of removing `std::enable_if` and adding `requires` in `/doxygen_cxx/classdrake_1_1solvers_1_1_mathematical_program.html`:
Before:
<img width="1848" height="802" alt="Image" src="https://github.com/user-attachments/assets/05d0407b-f7b4-46c1-8c5d-1655a43d0920" />

After:
<img width="1848" height="802" alt="Image" src="https://github.com/user-attachments/assets/6d2c5caf-bfdd-420b-ab2e-1c3bea43ea42" />

The result is that function signatures get shifted to the left. However, `std::tuple` return types (like `std::tuple< symbolic::Polynomial, MatrixXDecisionVariable, MatrixXDecisionVariable>` ) are mostly long which prevents function signatures from shifting more to the left. Should we alias the tuples or the result in the _after_ picture is satisfactory?

**Comment by jwnimmer-tri** [comment]
Well, "after" is certainly superior to "before", so we could accept that as as PR with nice progress.

For the tuples, I wonder if using `auto` as the return type would make things show up better?  And then explain the details of the `tuple` either as a ["trailing return type"](https://en.wikipedia.org/wiki/Trailing_return_type) (if Doxygen happens to show that better than a leading one), or just leave it as just "auto" and explain the details of the tuple only in the API documentation prose, not the function signature.

**Comment by j4yyousi** [comment]
Here's the improvement of the same section after implementing trailing return types:
<img width="1915" height="862" alt="Image" src="https://github.com/user-attachments/assets/b74a5907-53ff-41e6-94ec-d02a2954a64b" />

Doxygen does show trailing return types better than a leading one.

======================================================================
### Post processing method
## Background 
Doxygen's layout consists of a table with each function stored in a row `<tr>` html element. The return type of the function is stored in a `<td class="memItemLeft">` left cell, while the function name and its parameters are stored in a `<td class="memItemRight">` right cell. This structure cannot be changed by a config file, but rather by modifying Doxygen's source code. 
Too long return types cause inconveniences because: 
1.  all left cells have the same size as the longest left cell in the table,
1.  text in a left cell is displayed on a single line and doesn't wrap.
## Action
I have tried changing the layout of the page using Javascript. Here's the result:
<img width="1673" height="796" alt="Image" src="https://github.com/user-attachments/assets/90307ebe-3fd8-4866-a517-e98f5a8ee189" />
Pros: 
1. Row size is dynamic 
1. No need to modify header files.

Cons: layout is not as aesthetic as before.

Is this method worth pursuing?

**Comment by SeanCurtis-TRI** [comment]
Yeah. Making the cell widths dynamic has, in effect, eliminated the table-like nature. This feels like a step too far.

**Comment by jwnimmer-tri** [comment]
> Here's the improvement of the same section after implementing trailing return types:

Could you post a photo of what one of the trailing-return-type functions looks like now?  (Or just post a link to your branch and we can preview locally?)

> I have tried changing the layout of the page using Javascript.

If anything, changing the CSS would be better.  If there is a CSS way to say "this table column should never exceed 50% of the center div (or whole screen)", that might work okay.

But probably just landing the training return types is good enough for now.

**Comment by j4yyousi** [comment]
Here are the photos:
Before trailing return types:
<img width="1908" height="429" alt="Image" src="https://github.com/user-attachments/assets/a86a40dd-1cd0-4e57-b970-e890e12e8194" />

After trailing return types:
<img width="1914" height="317" alt="Image" src="https://github.com/user-attachments/assets/2aff2fa2-b317-4d38-9013-f94242130fb3" />

**Comment by j4yyousi** [comment]
Hi, regarding
>If anything, changing the CSS would be better. If there is a CSS way to say "this table column should never exceed 50% of the center div (or whole screen)", that might work okay.

I was able to achieve this at the expense of a scroll bar for the left cell as seen in the picture below:
<img width="1274" height="105" alt="Image" src="https://github.com/user-attachments/assets/bf0ed6ec-eb5f-4305-8446-ff283e20a896" />

I am assuming that wrapping text in the left cell is not allowed. Here's the CSS I added In `doxygen_extra.css`:
```
td.memItemLeft {
  white-space: nowrap;
  overflow-x: auto;
}

td.memItemRight{
  vertical-align: top;
}

table.memberdecls {
  border-collapse: collapse;
  table-layout: fixed;
  width: 100%;
}
```
I added this CSS to [the PR](https://github.com/RobotLocomotion/drake/pull/23808) as a general solution.
